### PR TITLE
Issue #4942: Updated javadoc grammar to make certain arguments mandatory for selected javadoc tags

### PR DESCRIPTION
**Pull request by ps-sp**
Issue #4942: Updated javadoc grammar to make certain arguments mandatory for selected javadoc tags

**NEED TO MERGE #5011 before looking into this.**

### CODE REVIEW

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 1]
```diff
@@ -861,19 +861,20 @@ reference:
...
 
-      | EXCEPTION_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) CLASS_NAME)? (WS | NEWLINE)*
-          ((WS | NEWLINE) description)?
+      | EXCEPTION_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+ CLASS_NAME
+          (WS | NEWLINE)* ((WS | NEWLINE) description)?
 
-      | PARAM_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) PARAMETER_NAME)? (WS | NEWLINE)*
-          ((WS | NEWLINE) description)?
+      | PARAM_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+ PARAMETER_NAME
+          (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Javadoc tool doesn't give complain about anything when description is missing but it does give a warning when no argument is supplied at all to `@param`. Therefore I feel that `PARAMETER_NAME` should be compulsory but `description` can be optional. This can be discussed further.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 2]
```diff
@@ -883,12 +884,12 @@ javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
       | SERIAL_FIELD_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) FIELD_NAME)? (WS | NEWLINE)*
           ((WS | NEWLINE) FIELD_TYPE)? (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: javadoc tool doesn't complain if no args are supplied to `@serial` `@serialData`  so I didn't touch them. 

`@serialField` behaves strangely as I wasn't able to get it to show on the site rendered by the javadoc tool even after sticking to the syntax and other requisities. **NOTE:** javadoc tool still gives a warning if no args are specified for `@serialField` tag, but since I couldn't get this tag to work properly _(or perhaps it was working properly and my expectations of it were wrong)_, I couldn't figure out what does javadoc tool treat the only argument as by default _i.e [does it treat it as `field-name` or `field-type` or `field-description`](http://docs.oracle.com/javase/7/docs/technotes/tools/windows/javadoc.html#serialField)_ . This needs to be looked into further.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 3]
```diff
@@ -883,12 +884,12 @@ javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
...
 
-      | THROWS_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) CLASS_NAME)? (WS | NEWLINE)*
-          ((WS | NEWLINE) description)?
+      | THROWS_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+ CLASS_NAME
+          (WS | NEWLINE)* ((WS | NEWLINE) description)?
 
-      | VERSION_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
+      | VERSION_LITERAL (WS | NEWLINE)+ description
 
       | CUSTOM_NAME (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: `description` should not be forced for custom tags I feel.

File: src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputSpaceBeforeDescriptionInBlockJavadocTags.javadoc [hunk 4]
```diff
@@ -38,16 +38,4 @@ include
...
-      Exception     Exception-description
-      spanning multiple lines
-
-* @param
-      ParamName
-      Param-description
-
-* @throws
-     Exception     Exception-description
-         spanning multiple lines
```
Comment (ps-sp) [inline src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputSpaceBeforeDescriptionInBlockJavadocTags.javadoc]: These input cases should be added in #4980 .

Comment (ps-sp) [reply src/test/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/javadocTags/InputSpaceBeforeDescriptionInBlockJavadocTags.javadoc]: This is because:

$ cat test.javadoc
```
* @exception
      Exception     Exception-description
      spanning multiple lines
```

java -jar master-8.2-all.jar -j test.javadoc
```
JAVADOC -> JAVADOC [0:0]
|--LEADING_ASTERISK -> * [0:0]
|--WS ->   [0:1]
|--JAVADOC_TAG -> JAVADOC_TAG [0:2]
|   |--EXCEPTION_LITERAL -> @exception [0:2]
|   |--NEWLINE -> \r\n [0:13]
|   |--WS ->        [1:0]
|   `--DESCRIPTION -> DESCRIPTION [1:6]
|       |--TEXT -> Exception     Exception-description [1:6]
|       |--NEWLINE -> \r\n [1:41]
|       `--TEXT ->       spanning multiple lines [2:0]
`--EOF -> <EOF> [2:29]
```

As can be seen `CLASS_NAME` which has been made mandatory for the corresponding javadoc tag is not being recognized in the above example and hence results in parse errors. This should be resolvable by adding `NEWLINE` to corresponding lexer modes but of course one can't be sure of it until a solution is attempted.

File: src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule713atclauses/InputNonEmptyAtclauseDescriptionCheck.java [hunk 5]
```diff
@@ -34,7 +34,7 @@ public int foo2(String a, int b, double c) throws Exception
 	 * @param a
 	 * @param b
 	 * @param c
-	 * @return
```
Comment (ps-sp) [inline src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule713atclauses/InputNonEmptyAtclauseDescriptionCheck.java]: `@return` should be removed from `NonEmptyAtclauseDescriptionCheck`. Also [`@exception`](http://docs.oracle.com/javase/7/docs/technotes/tools/windows/javadoc.html#exception) which is a synonym for `@throws` should be added to default tokens for the check since `@throws` is a default token for it. 

Also, there's a potential for other tags to be added to this check but that depends on the philosophy of the check. 

@romani @Vladlis @rnveach Please look into this. Should an issue be created for this

**Comment by ps-sp** [comment]
**NEED TO MERGE #5011 before looking into this.**

File: src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule713atclauses/InputNonEmptyAtclauseDescriptionCheck.java (hunk 5 above)
Comment (Vladlis) [reply src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule713atclauses/InputNonEmptyAtclauseDescriptionCheck.java]: yes, please open an issue

Comment (romani) [reply src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule713atclauses/InputNonEmptyAtclauseDescriptionCheck.java]: new issue should be created. 
It might be done on purpose previously as it was cleat that nobody in the works knows javadoc tag syntax.
as we become more strict in this area ... Checks should be updated, should be done completely separately after this PR is merged.

**Comment by romani** [comment]
@ps-sp , please rebase to resolve conflicts.
@Vladlis , please complete review.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 6]
```diff
@@ -877,20 +877,20 @@ reference:
       ;
 parameters: LEFT_BRACE (ARGUMENT | COMMA | WS | NEWLINE | LEADING_ASTERISK)* RIGHT_BRACE;
 
-javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
+javadocTag: AUTHOR_LITERAL (WS | NEWLINE)+ description
 
     | DEPRECATED_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: no complain from javadoc tool about `@deprecated` with no description

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 7]
```diff
@@ -877,20 +877,20 @@ reference:
...
+      | PARAM_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+ PARAMETER_NAME
+          (WS | NEWLINE)* ((WS | NEWLINE) description)?
 
-      | RETURN_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
+      | RETURN_LITERAL (WS | NEWLINE)+ description
 
-      | SEE_LITERAL (WS | NEWLINE)* reference? (STRING | htmlElement)* (WS | NEWLINE)*
-          ((WS | NEWLINE) description)?
+      | SEE_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+
+          (reference | STRING | htmlElement) (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Only one form of `@see` should be used at once and forms shouldn't be combined. That's why `(reference | STRING | htmlElement)` has been added

**Comment by ps-sp** [comment]
@romani 

>@ps-sp , please rebase to resolve conflicts.

Rebased.

@Vladlis 
Please have a look. I will redo the regression testing meanwhile.

**Comment by ps-sp** [comment]
[single javadoc check](https://4942-single-jcheck.netlify.com/)

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 (hunk 6 above)
Comment (Vladlis) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Ok

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 8]
```diff
@@ -877,20 +877,20 @@ reference:
...
 parameters: LEFT_BRACE (ARGUMENT | COMMA | WS | NEWLINE | LEADING_ASTERISK)* RIGHT_BRACE;
 
-javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
+javadocTag: AUTHOR_LITERAL (WS | NEWLINE)+ description
 
     | DEPRECATED_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
 
-      | EXCEPTION_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) CLASS_NAME)? (WS | NEWLINE)*
-          ((WS | NEWLINE) description)?
+      | EXCEPTION_LITERAL (WS | NEWLINE | {!isNextJavadocTag()}? LEADING_ASTERISK)+ CLASS_NAME
```
Comment (Vladlis) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Why add LEADING_ASTERISK here? Is there a corresponding tets case?
If LEADING_ASTERISK is needed I think it should be a separate issue.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 (hunk 7 above)
Comment (Vladlis) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: htmlElement allows any element while it should only be <a> tag, please open a new issue

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 (hunk 3 above)
Comment (Vladlis) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Ok

**Comment by romani** [comment]
we released 8.2 version, please rebase all your PRs to our latest master to avoid CI problems

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 (hunk 8 above)
Comment (ps-sp) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: Yes it is needed I will open an issue and add a test case, none verify this currently.
https://4942-single-jcheck.netlify.com/apache-ant/#A18

Comment (ps-sp) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: #5035

**Comment by ps-sp** [comment]
>htmlElement allows any element while it should only be tag, please open a new issue
#5036

**Comment by ps-sp** [comment]
@Vladlis @rnveach @romani 
Diffs look valid in the diff report. Please have a look.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 9]
```diff
@@ -877,20 +877,20 @@ reference:
       ;
 parameters: LEFT_BRACE (ARGUMENT | COMMA | WS | NEWLINE | LEADING_ASTERISK)* RIGHT_BRACE;
 
-javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
+javadocTag: AUTHOR_LITERAL (WS | NEWLINE)+ description
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: If there is a newline, doesn't it require a check for `LEADING_ASTERISK`?
Some of the others follow the same.

Comment (ps-sp) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: It will go into `description`.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 (hunk 2 above)
Comment (Vladlis) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: > This needs to be looked into further

Please open an issue

**Comment by rnveach** [comment]
@romani Please review and merge if you see no more issues.
